### DISCUSSION TITLE
Gama listening mode use

### DISCUSSION THREAD

**Discussion by lesquoyb** [gama-platform/gama#83] [Q&A] [answered]
There has been recently some commits by @hqnghi88 about gama "listening mode". This feature could be of great importance if working. So my questions are : is it really working ? If so: how do we use it ? Else can we help in some way ?

**Comment by hqnghi88** [comment, marked as answer]
HI,
It is a very fast devs, still needs to improve, if you want to test, you can use git release (alpha) . with the headless syntax:
gama-headless.bat -socket 6868, and use the web interface (file SimpleGUI.html) in msi.gama.headless plugin to interact with that listener

**Comment by RoiArthurB** [comment]
Everything works !! 🥳🥳🥳👏👏👏 
![image](https://user-images.githubusercontent.com/16764085/158739636-4ddcc273-7943-4040-98a9-0526dc7d366b.png)

Just displays, but it's probably normal for now 🤔

**Reply by hqnghi88** [reply]
No image from simulation?

**Reply by hqnghi88** [reply]
Try comokit lockdown ? Or even several connection session to one listeners?

**Reply by RoiArthurB** [reply]
Nop, no image what so ever ://

**Reply by hqnghi88** [reply]
ok i know, it s because that model has some else than displays (monitors) so i change a bit and reup for that case.

**Reply by hqnghi88** [reply]
<img width="912" alt="image" src="https://user-images.githubusercontent.com/6105384/158848111-69dd8f2f-aab6-4fea-adea-8316e4df6121.png">
should be ok now....

**Comment by hqnghi88** [comment]
@lesquoyb for the last point, helping is welcome. But for the moment there are some test i need to do so we will discuss soon.

**Reply by lesquoyb** [reply]
Sure, keep us in touch with your progress and what help you will need. We will need to document it at some point too

**Comment by hqnghi88** [comment]
Hi all,
I update some progress and some point need your supports.
First, this mode is base on websocket , so the communication must be done through websocket (R, C, GAML,...). 
The handshake connection mandatory on ip_address:port (ie, localhost:6868) , with the port is open by gama headless command: 
gama-headless.bat -socket 6868
gama-headless.sh -socket 6868

For the moment, here is the available endpoint (API):
localhost:6868/compile 
implemented to test a solution but did not commit, wonder about real use case. It serialize compiled model and send to client, server dont keep that object anymore (stateless)

localhost:6868/launch
stateful server api, accept different message tag (the [ ] is not in syntax, just to separate the idea):
launch@[model_path]@[experiment_name] 
compile and launch the experiment, return: exp@[id_exp]@[number_of_displays]

play@[experiment_id] 
play experiment (thread run), return at each simulation step: [serialized object of image or output values]

pause@[experiment_id] 
pause experiment, no return

step@[experiment_id] 
do one step , return [serialized object of outputs]

stop@[experiment_id] 
close experiment, no return

exit system exit 

localhost:6868/output
implemented to test a solution but did not commit, wonder about real use case, mostly for the dynamic dashboard, where customized evaluation/monitors can be add at runtime.
out@[number_of_displays] 
get a serialized object of output on demand, return [serialized object]

TODO in few days:
1.a solution to upload a zip from client machine, to server. The server accept this zip, then unzip in a temporary folder and compile/launch experiment.
2.standardize the serialized object (geojson or just string, bytearray) of an output on demand (agent population, with set of attributes...)

In the meantimes, to have this feature in official release 1.8.2, i need your support to test this mode and these api in some way:
1. a stand-alone js if you have skills, to not using my webgui, to see if the workflow can cause problem in communication with server??
2. in other language, platform, like R
3. Idea about infrastructure, structure are welcome
Thanks a lot.

**Comment by LucasGrjs** [comment]
Hello @hqnghi88,

Im looking to make gama communicate with many headless version of gama using java socket to make them work like master/slave,
I just want to use function to send/receive data, pause and step the simulation running on these headless.
Do you think it would work to use java socket with the same protocol of communication that your websocket is using to make it work ?

Thank you !

**Reply by hqnghi88** [reply]
Hi, Yes, it is supposed to be a "listener", it should communicate and be controlled by serial signal, tcp, udp socket , websocket or even restful api. It will be implemented in near future. If you could provide an simple example for your use case, it could help us to speedup the progess.
Thanks.

**Reply by LucasGrjs** [reply]
Hi, 

I was thinking of something like this for the master (I did wrote this example without using OutputStream and InputStream for socket to make it easy to write in but it would be the way to do it properly I think) :

```
public class Protocol {
	int idRequest; // 1 = pause, 2 = step, 3 = receive data, 4 = ask for data
        String data;
		
	public Protocol (int idRequest) {
		idRequest= idRequest;
	}
        public Protocol (int idRequest, String data) {
		idRequest= idRequest;
                data = data;
	}
}

void doPause(int socketID)
{
      getSocket(socketID).send(new Protocol(1));
}
void doStep(int socketID)
{  
      getSocket(socketID).send(new Protocol(2));
}
void sendData(int socketID, String data)
{  
      getSocket(socketID).send(new Protocol(3, data);
}
String getData(int SocketID)
{
      getSocket(socketID).send(new Protocol(4));
      getSocket(mySocket).recv(buffer);

     return buffer;
}
```

And so the headless version of this would be to process the request as you did for the WebSocket.
